plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+  ((k, a), Map k a)
+unsafeUncons m =
+  ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList tl)
+  where
+    (hd, tl) = P.unsafeUncons (toBuiltinList m)
+    (k, a) = P.pairToPair hd
+
+{-# INLINEABLE noDuplicateKeys #-}
+noDuplicateKeys :: forall k a. Map k a -> Bool
+noDuplicateKeys m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let k = BI.fst hd
+             in if member' k tl then False else go tl
+        )
+
+{-# INLINEABLE all #-}
+all :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
+all p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let a = P.unsafeFromBuiltinData (BI.snd hd)
+             in if p a then go tl else False
+        )
+
+{-# INLINEABLE any #-}
+any :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
+any p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        False
+        ( \hd tl ->
+            let a = P.unsafeFromBuiltinData (BI.snd hd)
+             in if p a then True else go tl
+        )
+
+{-# INLINEABLE union #-}
+
+-- | Combine two 'Map's.
+union ::
+  forall k a b.
+  (P.UnsafeFromData a, P.UnsafeFromData b, P.ToData a, P.ToData b) =>
+  Map k a ->
+  Map k b ->
+  Map k (These a b)
+union (toBuiltinList -> ls) (toBuiltinList -> rs) = unsafeFromBuiltinList res
+  where
+    ls' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    ls' = go ls
+      where
+        go xs =
+          P.matchList
+            xs
+            nil
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    v' = BI.snd hd
+                    v'' = case lookup' k' rs of
+                      Just r ->
+                        P.toBuiltinData
+                          ( These
+                              (P.unsafeFromBuiltinData v')
+                              (P.unsafeFromBuiltinData r) ::
+                              These a b
+                          )
+                      Nothing -> P.toBuiltinData (This (P.unsafeFromBuiltinData v') :: These a b)
+                 in BI.mkCons (BI.mkPairData k' v'') (go tl)
+            )
+
+    rs' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    rs' = go rs
+      where
+        go xs =
+          P.matchList
+            xs
+            nil
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    tl' = go tl
+                 in if member' k' ls
+                      then tl'
+                      else BI.mkCons hd tl'
+            )
+
+    res :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    res = go rs' ls'
```

Comment:
Can you make this act on `rs` and cons the missing stuff (ie the stuff not in the overlap) directly onto `ls'`? I think that at the moment it's using heap memory to construct `rs'` as an intermediate list which is then appended to `ls'` and discarded.  If you can avoid constructing `rs'` then it would be more efficient in terms of both CPU and memory.